racing_opt.jl

```diff
@@ -0,0 +1,213 @@
+    376, 374, 373, 372, 370, 369, 368, 367, 365, 364, 363, 362, 361, 360, 360, 359, 358, 357, 357, 356, 356, 355, 355,
+    354, 354, 354, 353, 353, 353, 353, 352, 352, 352, 352, 352, 351, 351, 351, 351, 351, 351, 351, 350, 350, 350, 350,
+    349, 349, 349, 348, 348, 347, 347, 346, 346, 345, 344, 344]
+
+x1_0 = 67    # Initial x position of player 1
+y1_0 = 343    # Initial y position of player 1
+v1_0 = 0.1    # Initial velocity of player 1
+theta1_0 = 1.25*pi    # Initial heading of player 1
+a1_0 = -2    # Initial acceleration of player 1
+phi1_0 = 0    # Initial steering angle of player 1
+tpx1 = 1 # Initial index of progression along track
+
+a_min1 = -6  # Maximum braking force for player 1
+a_max1 = 6  # Maximum acceleration for player 1
+phi_max1 = 30  # Maximum steering angle for player 1
+L1 = 0.35 # Length of player 1's car
+v_max1 = 5 # Max Velocity of player 1's car
+cornering_max1 = 1.0 # Max cornering Gs for player 1
+
+T = 3         # Time Horizon
+freq = 100    # Frequency
+n = T * freq  # Total Time steps
+
+@variables(car, begin
+    Δt ≥ 0, (start = 1 / n) # Time step
+    # State variables
+    0 ≤ v1[1:n] ≤ v_max1           # Velocity
+    -20 ≤ x1[1:n] ≤ 20    # X Position
+    -20 ≤ y1[1:n] ≤ 20    # Y Position
+    -20 ≤ theta1[1:n] ≤ 20    # Heading (I want to make this unbounded)
+
+    # Control variables
+    a_min1 ≤ a1[1:n] ≤ a_max1    # Acceleration
+    -phi_max1 ≤ phi1[1:n] ≤ phi_max1    # Acceleration
+end)
+
+
+function track_progression(x, y, idx, pos_horizon=30)
+ i = idx
+ min_dist = ((x-TRACK_CENTER_X[i])^2 + (y-TRACK_CENTER_Y[i])^2)
+ for j in idx:(idx + pos_horizon)
+    dist = ((x-TRACK_CENTER_X[j%length(TRACK_CENTER_X)])^2 + (y-TRACK_CENTER_Y[j%length(TRACK_CENTER_X)])^2)
+    if  min(dist, min_dist) == min_dist
+        min_dist = dist
+        i = j
+    end
+ end
+ return i, min_dist
+end
+
+
+fix(x1[1], x1_0; force = true)
+fix(y1[1], y1_0; force = true)
+fix(v1[1], v1_0; force = true)
+fix(theta1[1], theta1_0; force = true)
+fix(a1[1], a1_0; force = true)
+fix(phi1[1], phi1_0; force = true)
+
+@NLexpressions(
+    car,
+    begin
+        # Curvature(steering_angle) = tan(phi)/L
+        kappa1[j = 1:n], tan(phi1[j])/L1
+        # Time of travel
+        t_f, Δt * n
+    end
+)
+
+@objective(car, Max, sum(v1[j] * Δt for j in 1:n))
+
+
+function calculate_lateral_acceleration(v, kappa)
+    return (v^2)*kappa
+end
+
+for j in 2:n
+    # Dynamics Constraints
+    # Rectangular integration
+    @NLconstraint(car, x1[j] == x1[j - 1] + Δt * v1[j - 1]*cos(theta1[j - 1]))
+    @NLconstraint(car, y1[j] == y1[j - 1] + Δt * v1[j - 1]*sin(theta1[j - 1]))
+    @NLconstraint(car, theta1[j] == theta1[j - 1] + Δt * v1[j - 1]*tan(phi1[j - 1]))
+    @NLconstraint(car, v1[j] == v1[j - 1] + Δt * a1[j - 1])
+
+    # Track Limits Constraints (I am using logical operators in the following constraint, so it is not happy about this)
+    # @NLconstraint(car, track_progression(x1[j], y1[j], tpx1)[2] <= TRACK_WIDTH^2)
```

Comment:
@dfrt I read that there is no support for logical/conditional expressions in constraints and objectives. Do you have any suggestions on how to ensure a minimum distance is maintained to the center line of the track?